Test migration to VDR / Askar / CredX + DCO

VladimirWork opened the pull request:
Indy-node system tests migration from Indy SDK (libindy) to Indy VDR, Aries Askar and Indy CredX libraries.
Sovtoken-related tests are kept untouched since libsovtoken depends on libindy, not indy-vdr.
Test pipeline after all changes: https://github.com/VladimirWork/indy-test-automation/actions/runs/4135106879

WadeBarnes commented:
Apologies it's taken so long to get to this review.  I've got two instances of the tests running in a real world scenario.

- One run with payment tests **enabled**; https://github.com/WadeBarnes/sovrin/actions/runs/4254043535
- One run with payment tests **disabled**; https://github.com/WadeBarnes/sovrin/actions/runs/4254274241

The run with the payment tests **enabled** has encountered some test failures.  Would you be able to have a look?

The run with the payment tests **disabled** has only encountered one test failure, but that looks like a timeout.  I'll rerun the tests again once complete, I expect the test will pass when rerun.

I'm still in the process of reviewing the code.

WadeBarnes commented on system/indy-node-tests/TestLedgerSuite.py:
```diff
@@ -591,16 +604,17 @@ async def test_pool_restart(
     @pytest.mark.parametrize('reinstall', [False, True])
     @pytest.mark.parametrize('force', [False, True])
-    @pytest.mark.parametrize('package', ['indy-node', 'sovrin'])
+    # @pytest.mark.parametrize('package', ['indy-node', 'sovrin'])
+    @pytest.mark.parametrize('package', ['indy-node']) # unable to install sovrin package, it is installed via dpkg -i in Dockerfile
```
This should only be a problem is you are attempting to run the payment tests AND the Sovrin package is not installed.

WadeBarnes reviewed (changes requested):
I've reviewed the code.  Just a couple things noted below.

VladimirWork commented on system/indy-node-tests/TestMultiSigSuite.py:
```diff
@@ -111,41 +117,50 @@ async def test_case_steward_or_trust_anchor_can_send_nym(self, pool_handler, wal
+        # assert res_negative['op'] == 'REJECT'
+        req = ledger.build_nym_request(trustee_did, random_did_and_json()[0], None, None, None)
+        req = await multi_sign_request(wallet_handler, trustee_did, req)
+        with pytest.raises(indy_vdr.error.VdrError):
+            res_negative_multisign = await pool_handler.submit_request(req)
+        # assert res_negative_multisign['op'] == 'REJECT'
```
@MeltedGlaciers Let's proceed with the second option everywhere ^

VladimirWork commented:
@WadeBarnes Hi, PR is ready - we updated error handling and split tests and pipelines into 2 parts - new VDR-based and old token-related according to the selected option:

**Option 2: Keep sovrin token tests untouched and still depending on libsovtoken and libindy**
_We’ve decided that Option 2 is the best way to proceed.  It will be a Sovrin effort to migrate the payment related tests to use indy-vdr and to upgrade libsovtoken to use indy-vdr._

WadeBarnes commented on system_payments_only/README.md:
```diff
@@ -0,0 +1,12 @@
+# System Tests
```
```suggestion
# System Payments Only Tests

This directory contains tests that still depend on the indy-sdk to work.  These tests are related to the Sovrin packages including the token plugins which have yet to be migrated to use the newer shared libraries.
```

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -1,7 +1,7 @@
 on:
-  # push:
-  #   branches: [sovrin]
+  push:
+    branches: [main]
```
```
  push:
    branches: [main]
```
This should actually be removed.  The workflow should only be triggered through a workflow dispatch or workflow call.

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -137,6 +137,7 @@ env:
   NETWORK_SUBNET: ${{ fromJSON(inputs.networkJSON).networkSubnet }}
   TEST_DIRECTORY: ${{ inputs.testDirectory }}
   INCLUDE_PAYMENT_TESTS: ${{ inputs.includePaymentTests }}
+  TEST_DIRECTORY_PAYMENT: "system_payments_only/indy-node-tests" # keep sovrin token tests untouched and still depending on libsovtoken and libindy
```
The value for `TEST_DIRECTORY_PAYMENT` should be coming from an input so it can be overridden like `TEST_DIRECTORY`.

VladimirWork replied:
We can move both paths into TEST_DIRECTORY by switch it from string to key-value structure but it can broke your initial pipeline.

WadeBarnes replied:
Ok, we'll look into fixing that later.  @pSchlarb is working on reducing the number of required inputs.

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -225,18 +223,17 @@ jobs:
         uses: actions/checkout@v3
         with:
-          repository: hyperledger/indy-test-automation
+          repository: indy/indy-test-automation
```
```suggestion
          repository: hyperledger/indy-test-automation
```
It looks like a find and replace mistake, as it affects all occurrences.

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -261,11 +257,11 @@ jobs:
-        with:
-          name: sovrin-deb
-          path: ./system/docker/node/
+     - name: Download Pipeline sovrin.deb artifact
+       uses: actions/download-artifact@v3
+       with:
+         name: sovrin.deb
+         path: ./system/docker/node/
```
```suggestion
     - name: Download Pipeline sovrin-deb artifact
       uses: actions/download-artifact@v3
       with:
         name: sovrin-deb
         path: ./system/docker/node/
```
Otherwise the download of the artifact won't work.

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -405,13 +398,22 @@ jobs:
           # it clear the files are being ignored.
           # ==============================================
-          ignore_files_starts_with: ${{ inputs.ignoreTests }}
+          ignore_files_starts_with: ""
```
```suggestion
          ignore_files_starts_with: ${{ inputs.ignoreTests }}
```

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -405,13 +398,22 @@ jobs:
+          directory: ${{ env.TEST_DIRECTORY_PAYMENT }}
+          file_extension: "py"
+          only_files_starts_with: "test,Test"
+          ignore_files_starts_with: ""
```
```suggestion
          ignore_files_starts_with: ${{ inputs.ignoreTests }}
```

WadeBarnes reviewed (changes requested):
Looking good.  Just a few issues to note.

- Please remove the `sovrin_1.2.0_amd64.deb` files that got included by mistake.
- `matrix_test_action.yaml` looks to be unused.  If so, please delete.

I've re-run the test workflows with these latest updates:

- The run with payment tests **disabled** is working; https://github.com/WadeBarnes/sovrin/actions/runs/4254274241
  - I need to review the results but it looks good.

- The run with payment tests **enabled** is failing; https://github.com/WadeBarnes/sovrin/actions/runs/4254043535
  - I need to look into the issues.  Strange it's failing seeing as it's a relatively unchanged version of what's on the main branch.

VladimirWork commented:
@WadeBarnes Regarding payment pipeline failure:
Dockerfile.ubuntu-2004:110
--------------------
 109 |     RUN echo "SOVRIN_INSTALL: ${SOVRIN_INSTALL}"
 110 | >>> RUN if [ "${SOVRIN_INSTALL}" = "yes" ]; then \
 111 | >>>         sovrin_package=$(find ./ -name "sovrin_*.deb" | head -n 1); \
 112 | >>>         echo "Installing ${sovrin_package} ..."; \
 113 | >>>         dpkg -i "${sovrin_package}"; \
 114 | >>>     fi
 115 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c if [ \"${SOVRIN_INSTALL}\" = \"yes\" ]; then         sovrin_package=$(find ./ -name \"sovrin_*.deb\" | head -n 1);         echo \"Installing ${sovrin_package} ...\";         dpkg -i \"${sovrin_package}\";     fi" did not complete successfully: exit code: 1

Please recheck payment pipeline with the latest changes from this PR.

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -225,15 +221,14 @@ jobs:
       GITHUB_REPOSITORY_NAME: ${{ needs.workflow-setup.outputs.GITHUB_REPOSITORY_NAME }}
     steps:
-      - name: Git checkout hyperledger/indy-test-automation
+      - name: Git checkout indy/indy-test-automation
```
```suggestion
      - name: Git checkout hyperledger/indy-test-automation
```

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -261,11 +255,11 @@ jobs:
-        with:
-          name: sovrin-deb
-          path: ./system/docker/node/
+     - name: Download Pipeline sovrin-deb artifact
+       uses: actions/download-artifact@v3
+       with:
+         name: sovrin-deb
+         path: ./system/docker/node/
```
```suggestion
       - name: Download Pipeline sovrin-deb artifact
         uses: actions/download-artifact@v3
         with:
           name: sovrin-deb
           path: ./system/docker/node/
```
Fix indenting causing invalid workflow error.

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -372,17 +363,17 @@ jobs:
         uses: actions/checkout@v3
         with:
-          repository: hyperledger/indy-test-automation
+          repository: indy/indy-test-automation
```
```suggestion
          repository: hyperledger/indy-test-automation
```

WadeBarnes commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -446,29 +445,92 @@ jobs:
+      - name: Git checkout indy/indy-test-automation
+        uses: actions/checkout@v3
+        with:
+          repository: indy/indy-test-automation
```
```suggestion
          repository: hyperledger/indy-test-automation
```

WadeBarnes reviewed (changes requested):
Some additional adjustments to the workflow are needed.

WadeBarnes reviewed (changes requested):
Looking good.

The only issues I'm seeing now are with the test runs with the payment tests **enabled**; https://github.com/WadeBarnes/sovrin/actions/runs/4254043535.  I've reviewed the results from the test run with the payment tests **disabled** (https://github.com/WadeBarnes/sovrin/actions/runs/4254274241) and everything looks fine.

Ignoring the untouched `system_payments_only` tests that are in the default ignore list (`TestEndorserSuite,TestLedgerSuite,TestCatchUpSuite`) that are failing (for example the failure in `TestEndorserSuite` is a known issue), there are some additional tests that are failing that I'm expecting to be passing.

For example if you compare https://github.com/WadeBarnes/sovrin/actions/runs/4254043535 with https://github.com/sovrin-foundation/sovrin/actions/runs/4001099737, you will see there is and expected failure in the `test_state_proof` tests.

I'm rerunning the tests, but you can look at the run from Attempt 11 for results; https://github.com/WadeBarnes/sovrin/actions/runs/4254043535/attempts/11

pSchlarb commented on .github/workflows/callable_test-automation.yaml:
```diff
@@ -372,17 +363,17 @@ jobs:
+      payment-modules: ${{ steps.get-payment-module-list.outputs.file_names }}
     steps:
-      - name: Git checkout hyperledger/indy-test-automation
+      - name: Git checkout VladimirWork/indy-test-automation
```
Also noticed that this PR is still pointing to the GHA on your fork. 
While in development and finalizing the PR this is fine, but should be taken care of before the merge.

VladimirWork replied:
This is for testing purpose only, I will roll back this right before merging.

pSchlarb commented:
I am also not sure how these changes to the pipeline affect or conflict witht the changes from 
https://github.com/hyperledger/indy-test-automation/pull/126 and
https://github.com/sovrin-foundation/sovrin/pull/348

And the upcomming node integration.
Sorry i hadn't had the time to check for it. I hope i get this week some time to check it.

VladimirWork commented:
@WadeBarnes I've forked Sovrin main and run tests from [VladimirWork:vdr-askar-migration-copy] (https://github.com/VladimirWork/indy-test-automation/tree/vdr-askar-migration-copy) with payments, here is the result: https://github.com/VladimirWork/sovrin/actions/runs/4391344953 with 2 failures: TestEndorserSuite (as expected) and TestAdHocSuite with some strange exit code - looks like GHA runner issues, I'll rerun this job, but it looks like it was intermittent failure that doesn't relate to test code or env scripts in the indy test automation repo.

WadeBarnes commented:
Thanks @VladimirWork.  I'm rerunning things here to confirm the results; https://github.com/WadeBarnes/sovrin/actions/runs/4406874031

VladimirWork commented:
@WadeBarnes Your pipeline also looks good, I fixed conflicts with main and switched jobs back to hyperledger repo.

WadeBarnes commented:
@VladimirWork, Test results confirmed.  Results as expected.  Do you want to squash any of the commit history to clean things up a bit before I merge?

VladimirWork commented:
@WadeBarnes I squashed some unnecessary commits, I believe now PR is ready for merging.
